### PR TITLE
Segwit

### PR DESCRIPTION
**Pull request by lawrence860104**

### CODE REVIEW

**Comment by schildbach** [comment]
The branch at least needs a rebase against master. It won't be trivial. A lot of the stuff in this branch is already on master; not always in the same implementation.

File: core/src/main/java/org/bitcoinj/script/Script.java [hunk 1]
```diff
@@ -1755,4 +1904,66 @@ public boolean equals(Object o) {
...
+        if (isSentToP2WPKH()) {
+            byte[] pubKeyHash = Utils.sha256hash160(pubKeyOrScript);
+            if (witness.getPushCount() == 2 && Arrays.equals(expectedHash, pubKeyHash))
+                return scriptCode();
+            else
+                throw new ScriptException(ScriptError.SCRIPT_ERR_WITNESS_UNEXPECTED, "Incorrect public key hash or P2WPKH witness push count");
+        } else if (isSentToP2WSH()) {
+            byte[] scriptHash = Sha256Hash.hash(pubKeyOrScript);
+            if (Arrays.equals(expectedHash, scriptHash))
+                return new Script(pubKeyOrScript);
```
Comment (sbwdlihao) [inline core/src/main/java/org/bitcoinj/script/Script.java]: ByteArrayOutputStream stream = new ByteArrayOutputStream();
stream.write(new VarInt(pubKeyOrScript.length).encode());
stream.write(pubKeyOrScript);
return new Script(stream.toByteArray());

**Comment by jonathancross** [comment]
FTI: https://github.com/bitcoinj/bitcoinj/pull/1563 has been merged.

**Comment by schildbach** [comment]
Apart from the full verification, this should now be on master.

**Pull request closed without merging**